File: android/app/src/main/kotlin/net/guacamaya/proto/Crc16.kt
package net.guacamaya.proto

/**
 * CRC16-CCITT (poly 0x1021, init 0xFFFF, no reflection, no final XOR).
 *
 * Used as a cheap reject-before-sig-verify check on the 22-byte payload. NOT a
 * security primitive — only an optimization to drop junk frames before the
 * expensive Ed25519 verify.
 */
object Crc16 {

    fun ccitt(data: ByteArray, offset: Int = 0, length: Int = data.size): Int {
        var crc = 0xFFFF
        for (i in offset until offset + length) {
            crc = crc xor ((data[i].toInt() and 0xFF) shl 8)
            crc = crc and 0xFFFF
            repeat(8) {
                crc = if (crc and 0x8000 != 0) {
                    (crc shl 1) xor 0x1021
                } else {
                    crc shl 1
                }
                crc = crc and 0xFFFF
            }
        }
        return crc
    }
}

File: packages/shared/src/mesh/crc16.ts
/**
 * CRC16-CCITT (poly 0x1021, init 0xFFFF, no reflection, no final XOR).
 * Byte-for-byte mirror of net.guacamaya.proto.Crc16.ccitt (Android).
 */
export function crc16Ccitt(data: Uint8Array, offset = 0, length = data.length): number {
  let crc = 0xffff;
  for (let i = offset; i < offset + length; i++) {
    crc ^= (data[i]! & 0xff) << 8;
    crc &= 0xffff;
    for (let b = 0; b < 8; b++) {
      crc = (crc & 0x8000) !== 0 ? (crc << 1) ^ 0x1021 : crc << 1;
      crc &= 0xffff;
    }
  }
  return crc;
}
